### DISCUSSION TITLE
Using cuda-quantum with PyCharm

### DISCUSSION THREAD

**Discussion by poojarao8** [NVIDIA/cuda-quantum#1600] [Q&A]
@zivchen9993 is attempting to use the cuda-quantum docker container as the interpreter for PyCharm and are running into errors.  Pasting the error messages here: 

```
Also, I ran the code of the intro with the cudaq.set_target('nvidia') addition and I get two different results depending on where I run from:

If I run on the PyCharm with the cuda-quantum docker image as the interpreter, I receive this error upon running the cudaq.set_target('nvidia') line:
RuntimeError: [custatevec] %CUDA driver version is insufficient for CUDA runtime version in CuStateVecCircuitSimulator (line 334)

If I run on directly on the docker container and paste the same code, the code manages to run the cudaq.set_target('nvidia') line, but running:
@cudaq.kernel
def kernel(qubit_count: int):...           
generates this error:
OSError: could not get source code
```

@zivchen9993 is using WSL 2 with Ubuntu 22.04 and a Windows 64 bit machine and is pulling the nightly image.

Any guidance on next debugging steps or possible causes for the error would be great.  Thanks!

**Comment by bmhowe23** [comment]
Does the `nvidia-smi` command work inside the container? If not, is Nvidia Container Toolkit installed on the host (in WSL), and is the container started with `--gpus all`?

**Reply by zivchen9993** [reply]
thanks for the reply!
the `nvidia-smi` command works

**Reply by poojarao8** [reply]
Hey @zivchen9993, has anything changed on this?

**Reply by zivchen9993** [reply]
Hi Pooja, Sorry but not yet...

**Reply by bmhowe23** [reply]
Hi @zivchen9993 - I suspect the underlying cause of your two issues are different.

For the first issue (`PyCharm with the cuda-quantum docker image as the interpreter`), I _suspect_ PyCharm is not starting the container using the appropriate GPU resources. Could you please run these Python commands in that environment and verify that `nvidia-smi` is working in that environment?

```python
import subprocess
subprocess.run(['nvidia-smi'])
```

For the second issue (pasting code into an interactive Python session leads to `OSError: could not get source code`), that looks like a limitation of our kernel mode when using an interactive interpreter. I believe your options are to either:

1. Setup your PyCharm environment to use Python notebooks (i.e. `.ipynb` files instead of regular `.py` files), or
2. Avoid the `@cudaq.kernel` decorator syntax and use `cudaq.make_kernel()` instead (as shown [here](https://github.com/NVIDIA/cuda-quantum/blob/a1fad14061ff4191b06a8b36b897fe4d0e91016e/python/tests/builder/test_sample.py#L20))

(Full disclosure - I am not very familiar with PyCharm, so I don't know how feasible my .ipynb suggestion really is.)

**Reply by zivchen9993** [reply]
Hi @bmhowe23, Thanks a lot for the reply!

I've checked and regarding the first issue, you are right, the PyCharm does have a problem with it and it seems that I'm not the only one frustrated by it (and it doesn't seem like there's a solution). 
do you know of any nice IDE that won't have this kind of problem?

You were right about the second issue as well, I managed to run directly from the docker container and saw that my GPU was utilized during the process.

Thanks a lot and have a great weekend!

**Reply by bmhowe23** [reply]
> do you know of any nice IDE that won't have this kind of problem?

Many of us use [VS Code using these instructions](https://nvidia.github.io/cuda-quantum/latest/using/install/local_installation.html#development-with-vs-code). You can install many VS Code extensions to customize your Python experience. Hopefully this helps!
